### PR TITLE
Get the result of exec command

### PR DESCRIPTION
**Pull request by utam0k**
Fix: https://github.com/containers/youki/issues/953

<a href="https://gitpod.io/#https://github.com/containers/youki/pull/1018"></a>

### CODE REVIEW

**Comment by YJDoc2** [comment]
Hey @utam0k , sorry I haven't been much active here recently :sweat_smile: Let me know if you need any help with this!

**Comment by utam0k** [comment]
@YJDoc2 Thanks for your help. Can I ask you to help me with this pr? I am currently having trouble getting fifo's fd to pass to the init process, and I plan to get the results of exec's command execution through fifo.

**Comment by YJDoc2** [comment]
Hey, sure, I'll take a look and see if I can do anything by this weekend :)

**Comment by YJDoc2** [comment]
@utam0k should we make this ready for review and get review?
The conflict is minor and can be easily resolved

**Comment by utam0k** [comment]
@Furisto @yihuaf Can I ask you to review this PR ;)

File: crates/libcontainer/src/process/fork.rs [hunk 1]
```diff
@@ -8,15 +8,16 @@ use nix::unistd::Pid;
...
-                log::debug!("failed to run fork: {:?}", error);
-                -1
-            } else {
-                0
+            let ret = match cb() {
+                Err(error) => {
+                    log::debug!("failed to run fork: {:?}", error);
+                    -1
+                }
+                Ok(ec) => ec,
```
Comment (yihuaf) [inline crates/libcontainer/src/process/fork.rs:20]: Nit: Can we use good names? Such as `error_code`?

File: crates/libcontainer/src/process/fork.rs [hunk 2]
```diff
@@ -8,15 +8,16 @@ use nix::unistd::Pid;
 // using clone, we would have to manually make sure all the variables are
 // correctly send to the new process, especially Rust borrow checker will be a
 // lot of hassel to deal with every details.
-pub fn container_fork<F: FnOnce() -> Result<()>>(cb: F) -> Result<Pid> {
+pub fn container_fork<F: FnOnce() -> Result<i32>>(cb: F) -> Result<Pid> {
```
Comment (yihuaf) [inline crates/libcontainer/src/process/fork.rs:11]: Nit: I would suggest adding a bit of comment explaining the requirement for returning an error code. The `i32` unfortunately will take some time to decipher if people are not familiar with the code.

File: crates/libcontainer/src/process/container_main_process.rs [hunk 3]
```diff
@@ -23,12 +26,18 @@ pub fn container_main_process(container_args: &ContainerArgs) -> Result<Pid> {
     let init_chan = &mut channel::init_channel()?;
 
     let intermediate_pid = fork::container_fork(|| {
-        container_intermediate_process::container_intermediate_process(
+        let t = container_intermediate_process::container_intermediate_process(
```
Comment (yihuaf) [inline crates/libcontainer/src/process/container_main_process.rs:29]: Nit: can we have a better name than just `t`? Also maybe consider inline this with the match statement below.

File: crates/libcontainer/src/container/builder_impl.rs [hunk 4]
```diff
@@ -121,7 +122,8 @@ impl<'a> ContainerBuilderImpl<'a> {
             cgroup_manager: cmanager,
         };
 
-        let init_pid = process::container_main_process::container_main_process(&container_args)?;
+        let (intermediate, init_pid) =
```
Comment (yihuaf) [inline crates/libcontainer/src/container/builder_impl.rs:125]: Maybe I missed this, but can you add a bit of comment on the need to return the `intermediate` pid as well?

**Review by yihuaf — APPROVED** [review]
Some nits and requests for comments. Otherwise, LGTM.

**Comment by YJDoc2** [comment]
Hey @yihuaf , thanks for the review!
@utam0k all of these are from my changes, so I'll fix and open another PR at your branch

I'll make the changes as well as reply here, to keep the track.

File: crates/libcontainer/src/process/container_main_process.rs (hunk 3 above)
Comment (YJDoc2) [reply crates/libcontainer/src/process/container_main_process.rs:29]: :+1: inlined

File: crates/libcontainer/src/container/builder_impl.rs (hunk 4 above)
Comment (YJDoc2) [reply crates/libcontainer/src/container/builder_impl.rs:125]: can you take a look at this https://github.com/utam0k/youki/pull/115/files#diff-4c3bee02d644d36778359cdb85ba23c194655461d95e34c298659d4633c2459eR101-R106

File: crates/libcontainer/src/process/fork.rs (hunk 2 above)
Comment (YJDoc2) [reply crates/libcontainer/src/process/fork.rs:11]: can you take a look at this https://github.com/utam0k/youki/pull/115/files#diff-b0f79d7a9d35c79d1dfa4d453d78572414d768506564d63a2504d723e69b43b5R12-R14

File: crates/libcontainer/src/process/container_intermediate_process.rs [hunk 5]
```diff
@@ -115,7 +118,13 @@ pub fn container_intermediate_process(
         .close()
         .context("failed to close unused init sender")?;
 
-    Ok(())
+    if wait {
```
Comment (Furisto) [inline crates/libcontainer/src/process/container_intermediate_process.rs:121]: Can we just return the pid here and then do the wait in container_main_process (line 36)? I would like to handle this as high up as possible and leave the lower layers untouched.

Comment (YJDoc2) [reply crates/libcontainer/src/process/container_intermediate_process.rs:121]: Hey, @Furisto that was the original way I tried to do it, but the issue with it is that the final process (container process) is child of the intermediate process, so only that can wait for it. Trying to `waitpid` on the final process from the main process fails, as it is not main process's child. Thus we need to make intermediate wait for the final process, and main process wait for intermediate

Comment (Furisto) [reply crates/libcontainer/src/process/container_intermediate_process.rs:121]: I understand that. What I am asking is to do the wait [here](https://github.com/containers/youki/blob/0f3d89f5e70f68c8491aa84e9d2278fb15c7f306/crates/libcontainer/src/process/container_main_process.rs#L36), that's in the main_process file, but the code in that location is executed by the intermediate process.

Comment (YJDoc2) [reply crates/libcontainer/src/process/container_intermediate_process.rs:121]: Oh, ok, sorry I misunderstood earlier. Can you take a look at https://github.com/utam0k/youki/pull/116 and see if that is what you expected?

**Comment by YJDoc2** [comment]
Hey, there's a clippy lint failing, I'll fix that after @Furisto has seen the PR. That way if there are any other changes, I can make them as well.

**Review by Furisto — APPROVED** [review]

**Comment by utam0k** [comment]
Thanks!! @YJDoc2

**Comment by utam0k** [comment]
@YJDoc2 Have you tried the integration test of containerd yet?

**Comment by YJDoc2** [comment]
@utam0k  Unfortunately I haven't tried that yet, and am quite busy until this weekend, so probably will do it after that. Apologies for the delay. Also thanks a lot for helping with this PR! :smile:

**Comment by utam0k** [comment]
@YJDoc2 We passed :100: 
```
cat result.txt | grep TestContainerExecNoBinaryExists
=== RUN   TestContainerExecNoBinaryExists
=== PAUSE TestContainerExecNoBinaryExists
=== CONT  TestContainerExecNoBinaryExists
--- PASS: TestContainerExecNoBinaryExists (0.27s)
```

**Pull request merged**